File: render.py
#!/usr/bin/env python

import time

from datetime import datetime

import feedparser
import yaml

from jinja2 import Template


def chunk_list(lst, rows):
    return list((lst[i:i+rows] for i in range(0, len(lst), rows)))


def new_post(post_time, new_post_age_threshold):
    if post_time is None:
        parsed_post_time = time.mktime(datetime.now().timetuple())
    else:
        parsed_post_time = time.mktime(post_time)

    current_time = time.mktime(datetime.now().timetuple())
    if abs(current_time - parsed_post_time) < new_post_age_threshold:
        return True
    else:
        return False


def get_feeds(config):
    feeds = chunk_list(config['feeds'], config['rows'])

    for group in feeds:
        for feed in group:
            feed['posts'] = []
            feed_data = feedparser.parse(feed['rss_url'])
            for entry in feed_data.entries[:config['posts']]:
                new = new_post(
                    entry.get('published_parsed'),
                    config['new_post_age_threshold']
                )
                feed['posts'].append({
                    'title': entry['title'],
                    'link': entry['link'],
                    'new': new
                })
            print(f"added {len(feed['posts'])} posts for {feed['name']}")

    return feeds


def render_template(data):
    template = Template(open('template.html.j2').read())
    with open('index.html', 'w') as f:
        f.write(template.render(data=data))


def main():
    with open('config.yaml', 'r') as stream:
        config = yaml.safe_load(stream)

    data = {}
    data['title'] = config['site_title']
    data['feeds'] = get_feeds(config)
    data['extra_links'] = chunk_list(config['extra_links'], config['rows'])

    render_template(data)


if __name__ == "__main__":
    main()

File: render.rb
#!/usr/bin/env ruby

require 'cgi'
require 'date'
require 'erb'
require 'net/http'
require 'open-uri'
require 'rss'
require 'time'
require 'yaml'

$stdout.sync = true

def get_feed(feed_url, timeout)
  3.times do
    begin
      feed_content = nil

      URI.open(feed_url, read_timeout: timeout, open_timeout: timeout) do |f|
        feed_content = f.read
      end

      feed = RSS::Parser.parse(feed_content, false)
      return feed if feed && feed.items.any?
    rescue => e
      puts "Error while fetching feed: #{feed_url}: #{e}"
      next
    end
  end

  nil
end

data = YAML.load_file('feeds.yaml')

data['feeds'].each do |name, info|
  info['posts'] = []
  feed_data = get_feed(info['feed'], info['timeout'] || 10)

  if feed_data.nil?
    data['feeds'].delete(name)
    next
  end

  feed_data.items.each do |entry|
    next if info['posts'].any? { |post| post['title'] == entry.title }

    post_date = case entry
                when RSS::Atom::Feed::Entry
                  entry.published ? Time.parse(entry.published.to_s).utc : Time.now.utc
                else
                  entry.pubDate ? Time.parse(entry.pubDate.to_s).utc : Time.now.utc
                end

    info['posts'] << {
      'title' => CGI.unescapeHTML(entry.title.to_s.gsub(/<[^>]*>/, '')).strip,
      'link' => entry.link.respond_to?(:href) ? entry.link.href : entry.link.to_s,
      'date' => post_date,
      'new' => (Time.now.utc - post_date).abs < 86400
    }

    if info['posts'].length >= 5
      info['posts'].sort! { |a, b| b['date'] <=> a['date'] }
      break
    end
  end

  puts "posts captured for #{name}"
end

template = ERB.new(File.read('template.html.erb'))
html = template.result(binding)

Dir.mkdir('rendered') unless Dir.exist?('rendered')

File.write('rendered/index.html', html)
puts "template rendered"
